File: 2024/python/day6.py
from datetime import datetime
import multiprocessing

MAP = []
with open('2024/_input/day6.txt') as fp:
    for line in fp.readlines():
        MAP.append(list(line))

UP = (-1, 0)
DOWN = (1, 0)
RIGHT = (0, 1)
LEFT = (0, -1)

def rotate(d):
    if d == DOWN:
        return LEFT
    if d == LEFT:
        return UP
    if d == UP:
        return RIGHT
    if d == RIGHT:
        return DOWN
    return (0, 0)

def find_guard():
    for i, r in enumerate(MAP):
        for j, c in enumerate(r):
            if c == '^':
                return (i, j)
    return (-1, -1)

def on_map(x, y):
    return x >= 0 and y >= 0 and x < len(MAP) and y < len(MAP[0])

def get_char(x, y, block=(-1,-1)):
    if x == block[0] and y == block[1]:
        return '#'
    if not on_map(x, y):
        return None
    return MAP[x][y]

def part1():
    # initial direction
    dr, dc = UP
    # initial position
    r, c = find_guard()

    V = set()
    while on_map(r, c):
        # turn if we need to
        ch = get_char(r + dr, c + dc)
        while ch == '#':
            dr, dc = rotate((dr, dc))
            ch = get_char(r + dr, c + dc)

        # move by the movement delta
        r = r + dr
        c = c + dc

        # remember this location, distinctly
        V.add((r, c))

    return V

def is_loop(block):
    # initial direction
    dr, dc = UP
    # initial position
    r, c = find_guard()

    V = set()
    while on_map(r, c):

        # turn if we need to
        ch = get_char(r + dr, c + dc, block)
        while ch == '#':
            dr, dc = rotate((dr, dc))
            ch = get_char(r + dr, c + dc)

        # move by the movement delta
        r = r + dr
        c = c + dc

        # have we been here before?
        DC = (r, c, dr, dc)
        if DC in V:
            return True

        V.add(DC)

    return False

if __name__ == '__main__':
    start = datetime.now()
    path = part1()
    print('part 1:', len(path), 'steps')
    with multiprocessing.Pool(16) as pool:
        print('part 2:', sum([1 if r else 0 for r in pool.map(is_loop, path)]), 'possibilities')
        pool.close()
    print('elapsed:', datetime.now() - start)

File: 2024/go/day6.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"time"
)

var MAP [][]rune

var (
	UP    = Coord{-1, 0}
	DOWN  = Coord{1, 0}
	RIGHT = Coord{0, 1}
	LEFT  = Coord{0, -1}
)

type Coord struct {
	x, y int
}

type State struct {
	pos Coord
	dir Coord
}

func rotate(d Coord) Coord {
	switch d {
	case DOWN:
		return LEFT
	case LEFT:
		return UP
	case UP:
		return RIGHT
	case RIGHT:
		return DOWN
	}
	return Coord{}
}

func findGuard() Coord {
	for i := range MAP {
		for j := range MAP[i] {
			if MAP[i][j] == '^' {
				return Coord{i, j}
			}
		}
	}
	return Coord{}
}

func onMap(x, y int) bool {
	return x >= 0 && y >= 0 && x < len(MAP) && y < len(MAP[0])
}

func getChar(x, y int) rune {
	if !onMap(x, y) {
		return 0
	}
	return MAP[x][y]
}

func getCharWithBlock(x, y int, block Coord) rune {
	if x == block.x && y == block.y {
		return '#'
	}
	return getChar(x, y)
}

func part1() map[Coord]struct{} {
	// initial direction
	dir := UP
	// initial position
	pos := findGuard()
	visited := make(map[Coord]struct{})

	for onMap(pos.x, pos.y) {
		// turn if we need to
		ch := getChar(pos.x+dir.x, pos.y+dir.y)
		for ch == '#' {
			dir = rotate(dir)
			ch = getChar(pos.x+dir.x, pos.y+dir.y)
		}
		// move by the movement delta
		pos.x += dir.x
		pos.y += dir.y
		// remember this location, distinctly
		visited[pos] = struct{}{}
	}
	return visited
}

func isLoop(block Coord) bool {
	// initial direction
	dir := UP
	// initial position
	pos := findGuard()
	visited := make(map[State]struct{})

	for onMap(pos.x, pos.y) {
		// turn if we need to
		ch := getCharWithBlock(pos.x+dir.x, pos.y+dir.y, block)
		for ch == '#' {
			dir = rotate(dir)
			ch = getCharWithBlock(pos.x+dir.x, pos.y+dir.y, block)
		}
		// move by the movement delta
		pos.x += dir.x
		pos.y += dir.y

		// have we been here before with the same direction?
		state := State{pos, dir}
		_, exists := visited[state]
		if exists {
			return true
		}
		visited[state] = struct{}{}
	}
	return false
}

func main() {
	start := time.Now()

	// Read input file
	file, err := os.Open("2024/_input/day6.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		MAP = append(MAP, []rune(scanner.Text()))
	}
	path := part1()
	fmt.Printf("part 1: %d steps\n", len(path))

	count := 0
	for coord := range path {
		if isLoop(coord) {
			count++
		}
	}
	fmt.Printf("part 2: %d possibilities\n", count)
	fmt.Printf("elapsed: %v\n", time.Since(start))
}
